File: app/javascript/channels/lobby_channel.js
import consumer from "./consumer"

document.addEventListener('DOMContentLoaded', () => {
  const enterButton = document.getElementById('enter-lobby')
  const nameInput = document.getElementById('name-input')
  const lobbyDiv = document.getElementById('lobby')
  const nameFormDiv = document.getElementById('name-form')
  
  enterButton.addEventListener('click', () => {
    const name = nameInput.value
    if (name.trim() === "") {
      alert("Please enter your name")
      return
    }
    
    lobbyDiv.style.display = "block"
    nameFormDiv.style.display = "none"

    const gameCode = document.getElementById('game-code').value

    const subscription = consumer.subscriptions.create({ channel: "LobbyChannel", game_code: gameCode, name: name }, {
      connected() {
        // Called when the subscription is ready for use on the server
        console.log("connected")
      },

      disconnected() {
        // Called when the subscription has been terminated by the server
      },

      received(data) {
        if (data.users) {
          // Update user list
          const userList = document.getElementById('user-list')
          userList.innerHTML = ""
          data.users.forEach(user => {
            const userElement = document.createElement('div')
            userElement.innerText = user
            userList.appendChild(userElement)
          })
        }
      },

    });
  })
});

File: app/channels/lobby_channel.rb
class LobbyChannel < ApplicationCable::Channel
  def subscribed
    @game = Game.find_by(code: params[:game_code])
    if @game
      @game.add_player(params[:name])
      stream_from "lobby_channel_#{params[:game_code]}"
      broadcast_user_list
      ActionCable.server.broadcast "lobby_channel_#{params[:game_code]}", { message: "#{params[:name]} has joined the lobby" }
    else
      reject
    end
  end

  def unsubscribed
    @game = Game.find_by(code: params[:game_code])
    if @game
      @game.remove_player(params[:name])
      broadcast_user_list
      ActionCable.server.broadcast "lobby_channel_#{params[:game_code]}", { message: "#{params[:name]} has left the lobby" }
    end
  end
  private

  def broadcast_user_list
    ActionCable.server.broadcast "lobby_channel_#{params[:game_code]}", { users: @game.initial_state['players'] }
  end

end
